File: quadro/quadro/decoder.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from quadro_interface.srv import  Decoder


class decoder(Node):
    def __init__(self):
        super().__init__('decoder')
        self.subscription = self.create_subscription(String, 'key_data', self.callback, 10)
        self.key = ''
        self.decoderSrv = self.create_service(Decoder, 'decoder', self.decode)
    def callback(self, msg):
        self.key = msg.data
        self.get_logger().info('received decoder key : %s' %self.key)


    def decode(self, request, response):
        j = 0
        data = list(request.encoded)
        for i,item in enumerate(data):
            if i%2 == 0 and i<len(data)-1:
                temp = data[i+1]
                data[i+1]= data[i]
                data[i] = temp
        for i,item in enumerate(data):
            next_pos = ord(item) - ord(self.key[j])
            while next_pos <32:
                next_pos = -32+(next_pos+126)
            data[i] = chr(next_pos)
            j+=1
            if j == len(self.key):
                j = 0
        response.data = ''.join(data)[::-1]
        return response
    
def main(args = None):
    rclpy.init(args=args)
    cod = decoder()
    rclpy.spin(cod)
    cod.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: quadro_ext/src/decoder.cpp
#include "rclcpp/rclcpp.hpp"
#include "quadro_interface/srv/binary_decoder.hpp"   
#include "std_msgs/msg/string.hpp"
#include <memory>
#include <string>
#include <vector>

using std::placeholders::_1;
using std::placeholders::_2;
using std::placeholders::_3;
using std::string;
class Decoder : public rclcpp::Node{



public:
    Decoder()
    : Node("Decoder"){
         // CHANGE
      rclcpp::Service<quadro_interface::srv::BinaryDecoder>::SharedPtr service = 
      this->create_service<quadro_interface::srv::BinaryDecoder>("binary_decoder",  &Decoder::decode);
      subscriber_ = this->create_subscription<std_msgs::msg::String>("key_topic", 10, std::bind(&Decoder::topic_callback, this, _1));

    }

  private:
    string key_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscriber_;
    size_t count_;
    void decode(std::shared_ptr<quadro_interface::srv::BinaryDecoder::Request> request,
    std::shared_ptr<quadro_interface::srv::BinaryDecoder::Response> response){

                   }

    void topic_callback(const std_msgs::msg::String & msg)
    {
      key_ = msg.data;
    }
    
};















int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<Decoder>());
  rclcpp::shutdown();
  return 0;



}